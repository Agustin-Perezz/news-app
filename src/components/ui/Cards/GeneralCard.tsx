import { CardActionArea, CardContent, Typography, Box } from '@mui/material';
import { NewsProps } from '../../../types/news-props';
import { CategoriesArticles } from './Categories';
import { CustomCard, CustomChip } from './CustomStylesCard';
import { useLoadImage } from '../../../hooks';

import load_image from '../../../assets/images/load-image.jpg';

interface Props {
  dataArticle: NewsProps;
}

export const GeneralCard: React.FC<Props> = ({ dataArticle }) => {
  const { loaded, onError, onLoad } = useLoadImage();

  return (
    <CustomCard>
      <CardActionArea target="_blank" href={dataArticle.url} sx={{ height: '470px' }}>
        <Box sx={{ position: 'relative' }}>
          <img
            style={{ height: '185px' }}
            onLoad={onLoad}
            alt={dataArticle.nameSource}
            src={loaded ? dataArticle.image_url : load_image}
            onError={({ currentTarget }) => onError({ currentTarget })}
          />
          <CustomChip label={dataArticle.published_at} />
        </Box>
        <CardContent>
          <Typography data-testid="card-title" variant="subtitle1">
            {dataArticle.title}
          </Typography>
          <Typography data-testid="card-description" variant="body2" color="text.secondary">
            {dataArticle.description}
          </Typography>
          {dataArticle.categories && <CategoriesArticles listCategories={dataArticle.categories} />}
          {dataArticle.nameSource && (
            <Box sx={{ display: 'flex', marginTop: '10px' }}>
              <Typography variant="subtitle2" mr={1}>
                Source:
              </Typography>
              <Typography variant="body2" color="text.secondary" sx={{ lineHeight: '1.57' }}>
                {dataArticle.nameSource}
              </Typography>
            </Box>
          )}
        </CardContent>
      </CardActionArea>
    </CustomCard>
  );
};
