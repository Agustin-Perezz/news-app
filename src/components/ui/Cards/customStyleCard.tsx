import { styled, createTheme } from '@mui/material/styles';
import MuiCard from '@mui/material/Card';
import MuiChip from '@mui/material/Chip';

export const tabFont = createTheme({
  typography: {
    fontFamily: ['Heebo', 'sans-serif'].join(','),
  },
  palette: { secondary: { main: '#2475C7' } },
});

export const CustomCard = styled(MuiCard)(() => ({
  '&:hover': {
    boxShadow: '0px 2px 18px 0px rgba(0,0,0,0.8)',
    transform: 'scale(1.02)',
    transition: 'all .2s ease-in-out',
    cursor: 'pointer',
  },
  maxHeight: 'auto',
  minHeight: '447px',
  width: '278px',
}));

export const CustomChip = styled(MuiChip)(() => ({
  position: 'absolute',
  top: '8px',
  right: '5px',
  color: 'ghostwhite',
  backgroundColor: '#0A1929',
  opacity: '0.6',
  height: '29px',
  border: '1px solid #9393A5',
}));
