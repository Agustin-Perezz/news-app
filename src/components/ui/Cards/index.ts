export * from './ListCards';
export * from './customStyleCard';
