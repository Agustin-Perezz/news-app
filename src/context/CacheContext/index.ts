export * from './CacheContext';
export * from './CacheProvider';
