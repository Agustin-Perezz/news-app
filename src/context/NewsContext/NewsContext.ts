import { createContext } from 'react';
import { NewsProps } from '../../types/news-props';

export type ContextTypeData = {
  data?: NewsProps[];
  isLoading: boolean;
  isError?: string;
};

export type NewsContextProps = ContextTypeData & {
  setState: React.Dispatch<React.SetStateAction<ContextTypeData>>;
};

export const NewsContext = createContext<NewsContextProps>(
  {} as NewsContextProps
);
