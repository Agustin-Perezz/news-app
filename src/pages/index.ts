export * from './Categories';
export * from './Header';
export * from './Home';
export * from './News';
