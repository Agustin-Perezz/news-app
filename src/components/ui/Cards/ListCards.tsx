import React from 'react';
import { Grid } from '@mui/material';
import { NewsProps } from '../../../types/news-props';
import { GeneralCard } from './GeneralCard';
import { RenderController } from '../RenderController';
import { CustomGrid, CustomTitle } from './CustomStyleCard';

interface Props {
  dataArticles?: NewsProps[];
  isLoading: boolean;
  isError?: string;
  titleText?: string;
}

export const ListCards: React.FC<Props> = ({
  dataArticles,
  isLoading,
  isError,
  titleText,
}) => {
  return (
    <RenderController
      isLoading={isLoading}
      isError={isError}
      data={dataArticles}
    >
      <Grid sx={{ flexGrow: 1, margin: '25px 0px' }}>
        <Grid
          item
          xs={12}
          sx={{
            justifyContent: 'center',
            display: 'flex',
            position: 'relative',
          }}
        >
          <CustomGrid container spacing={3}>
            {dataArticles?.map((article, index) => (
              <Grid key={article.uuid} item>
                <GeneralCard dataArticle={article} />
                {index === 0 && (
                  <CustomTitle variant="h6"> {titleText} </CustomTitle>
                )}
              </Grid>
            ))}
          </CustomGrid>
        </Grid>
      </Grid>
    </RenderController>
  );
};
