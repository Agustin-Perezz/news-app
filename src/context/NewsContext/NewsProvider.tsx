import React from 'react';
import { createAddaptNew } from '../../adapters';
import { newsApi } from '../../api';
import { useAxios } from '../../hooks';
import { NewsProps, NewsResponseProps } from '../../types';
import { NewsContext } from './NewsContext';

type Props = {
  children: JSX.Element | JSX.Element[];
};

type StateProps = {
  newsData: NewsProps[] | null;
};

const API_KEY = process.env.REACT_APP_API_KEY;
const INIT_STATE: StateProps = { newsData: null };

export const NewsProvider: React.FC<Props> = ({ children }) => {
  const [state, setState] = React.useState(INIT_STATE);
  const { fetchData, isLoading, isError } = useAxios();

  const getNewsByQuery = async (query: string) => {
    const { data } = await fetchData<NewsResponseProps>({
      programApi: newsApi,
      url: `/all?api_token=${API_KEY}&search=${query}`,
    });
    const listNews = createAddaptNew(data);
    setState({ newsData: listNews });
  };

  return (
    <NewsContext.Provider
      // eslint-disable-next-line react/jsx-no-constructed-context-values
      value={{ ...state, isLoading, isError, getNewsByQuery }}
    >
      {children}
    </NewsContext.Provider>
  );
};
