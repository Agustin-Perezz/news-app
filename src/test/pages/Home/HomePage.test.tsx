import { screen } from '@testing-library/react';
import { NewsContext } from '../../../context/NewsContext';
import { HomePage } from '../../../pages';
import { mockArticles } from '../../components/ui/Cards/mockArticles';
import { renderWithCache } from '../../utils';

describe('Test in <HomePage />', () => {
  const setState = jest.fn();
  function setup() {
    const { container } = renderWithCache({
      cacheValue: { country: 'Argentina' },
      route: '/home',
      children: (
        // eslint-disable-next-line react/jsx-no-constructed-context-values
        <NewsContext.Provider value={{ isLoading: false, setState, data: mockArticles }}>
          <HomePage />
        </NewsContext.Provider>
      ),
    });
    return container;
  }

  it('should match snapshot', () => {
    const container = setup();
    expect(container).toMatchSnapshot();
  });
  it('should show initial messages', () => {
    setup();
    screen.getByText('Welcome.');
    screen.getByText('Top news from Argentina.');
  });
  it('should show defaults top news in view', async () => {
    const container = setup();
    mockArticles.forEach((article) => {
      expect(screen.getByText(article.title)).toBeInTheDocument();
    });
    const cards = container.getElementsByClassName('MuiGrid-item');
    expect(cards.length).toBe(3);
  });
});
