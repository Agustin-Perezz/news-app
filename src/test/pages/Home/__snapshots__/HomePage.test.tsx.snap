// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <HomePage /> should match snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-14hrs0g"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4 css-1r4f7b3-MuiTypography-root"
    >
      Welcome.
    </h4>
    <div>
      <div
        class="MuiGrid-root css-87ggms-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-1b9hrwo"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-qscxtw-MuiTypography-root"
          >
            Top news from Argentina.
          </p>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-z7x17r-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-bzgqu1-MuiGrid-root"
            style="opacity: 1; webkit-transition: opacity 1500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 1500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
          >
            <div
              class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
            >
              <div
                class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-nkszem-MuiPaper-root-MuiCard-root"
              >
                <a
                  class="MuiButtonBase-root MuiCardActionArea-root css-y29lt3-MuiButtonBase-root-MuiCardActionArea-root"
                  href="https://google.com"
                  tabindex="0"
                  target="_blank"
                >
                  <div
                    class="MuiBox-root css-79elbk"
                  >
                    <img
                      alt="calcionapoli24.it"
                      src="[object Object]"
                      style="height: 185px; min-width: 100%;"
                    />
                    <div
                      class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-c8zb3x-MuiChip-root"
                    >
                      <span
                        class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                      >
                        11/25/2022
                      </span>
                    </div>
                  </div>
                  <div
                    class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
                  >
                    <h6
                      class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
                      data-testid="card-title"
                    >
                      Omaggio Maradona a Fuorigrotta, c'Ã¨ anche Maradona junior | FOTO
                    </h6>
                    <p
                      class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
                      data-testid="card-description"
                    >
                      Anche il figlio del Pibe presente stasera a Fuorigrotte
                    </p>
                    <div
                      class="MuiBox-root css-3yoo1r"
                    >
                      <h6
                        class="MuiTypography-root MuiTypography-subtitle2 css-1ij1w03-MuiTypography-root"
                      >
                        Categories:
                      </h6>
                      <div
                        class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-s89mk0-MuiChip-root"
                      >
                        <span
                          class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                        >
                          tech
                        </span>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root css-1hprwxc"
                    >
                      <h6
                        class="MuiTypography-root MuiTypography-subtitle2 css-1od5i09-MuiTypography-root"
                      >
                        Source:
                      </h6>
                      <p
                        class="MuiTypography-root MuiTypography-body2 css-bssea4-MuiTypography-root"
                      >
                        calcionapoli24.it
                      </p>
                    </div>
                  </div>
                  <span
                    class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
                  />
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </a>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
            >
              <div
                class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-nkszem-MuiPaper-root-MuiCard-root"
              >
                <a
                  class="MuiButtonBase-root MuiCardActionArea-root css-y29lt3-MuiButtonBase-root-MuiCardActionArea-root"
                  href="https://google.com"
                  tabindex="0"
                  target="_blank"
                >
                  <div
                    class="MuiBox-root css-79elbk"
                  >
                    <img
                      alt="calcionapoli24.it"
                      src="[object Object]"
                      style="height: 185px; min-width: 100%;"
                    />
                    <div
                      class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-c8zb3x-MuiChip-root"
                    >
                      <span
                        class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                      >
                        11/25/2022
                      </span>
                    </div>
                  </div>
                  <div
                    class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
                  >
                    <h6
                      class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
                      data-testid="card-title"
                    >
                      Title test
                    </h6>
                    <p
                      class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
                      data-testid="card-description"
                    >
                      description test
                    </p>
                    <div
                      class="MuiBox-root css-3yoo1r"
                    >
                      <h6
                        class="MuiTypography-root MuiTypography-subtitle2 css-1ij1w03-MuiTypography-root"
                      >
                        Categories:
                      </h6>
                      <div
                        class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-s89mk0-MuiChip-root"
                      >
                        <span
                          class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                        >
                          sports
                        </span>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root css-1hprwxc"
                    >
                      <h6
                        class="MuiTypography-root MuiTypography-subtitle2 css-1od5i09-MuiTypography-root"
                      >
                        Source:
                      </h6>
                      <p
                        class="MuiTypography-root MuiTypography-body2 css-bssea4-MuiTypography-root"
                      >
                        calcionapoli24.it
                      </p>
                    </div>
                  </div>
                  <span
                    class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
                  />
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
